FeedStoreChallenge/Infrastructure/CoreData/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,54 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				if let managedCache = try ManagedCache.find(in: context) {
+					completion(.found(feed: managedCache.localFeed, timestamp: managedCache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				let cache = try ManagedCache.newUniqueInstance(context: context)
+				cache.timestamp = timestamp
+				cache.feed = ManagedFeedImage.images(from: feed, in: context)
+				try context.save()
+
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				try ManagedCache.find(in: context).map(context.delete)
+				if (context.hasChanges) {
```

Comment:
The parenthesis can be removed:

```suggestion
				if context.hasChanges {
```